File: JavaIntroNeuralNetworkEdition2/src/com/heatonresearch/book/introneuralnet/ch6/TicTacToe/players/neural/TicTacToeChromosome.java
/**
 * Introduction to Neural Networks with Java, 2nd Edition
 * Copyright 2008 by Heaton Research, Inc. 
 * http://www.heatonresearch.com/books/java-neural-2/
 * 
 * ISBN13: 978-1-60439-008-7  	 
 * ISBN:   1-60439-008-5
 *   
 * This class is released under the:
 * GNU Lesser General Public License (LGPL)
 * http://www.gnu.org/copyleft/lesser.html
 */
package com.heatonresearch.book.introneuralnet.ch6.TicTacToe.players.neural;

import com.heatonresearch.book.introneuralnet.ch6.TicTacToe.game.ScorePlayer;
import com.heatonresearch.book.introneuralnet.ch6.TicTacToe.players.Player;
import com.heatonresearch.book.introneuralnet.neural.exception.NeuralNetworkError;
import com.heatonresearch.book.introneuralnet.neural.feedforward.FeedforwardNetwork;
import com.heatonresearch.book.introneuralnet.neural.feedforward.train.genetic.NeuralChromosome;

/**
 * Chapter 6: Training using a Genetic Algorithm
 * 
 * TicTacToeChromosome: Implements a chromosome for the neural network
 * that plays tic-tac-toe.
 * 
 * @author Jeff Heaton
 * @version 2.1
 */
public class TicTacToeChromosome extends NeuralChromosome<TicTacToeGenetic> {

	/**
	 * The constructor, takes a list of cities to set the initial "genes" to.
	 * 
	 * @param cities
	 *            The order that this chromosome would visit the cities. These
	 *            cities can be thought of as the genes of this chromosome.
	 * @throws NeuralNetworkException
	 */
	public TicTacToeChromosome(final TicTacToeGenetic genetic,
			final FeedforwardNetwork network) throws NeuralNetworkError {
		this.setGeneticAlgorithm(genetic);
		this.setNetwork(network);

		initGenes(network.getWeightMatrixSize());
		updateGenes();
	}

	@Override
	public void calculateCost() {

		try {
			// update the network with the new gene values
			this.updateNetwork();

			final PlayerNeural player1 = new PlayerNeural(getNetwork());
			Player player2;

			player2 = (Player) this.getGeneticAlgorithm().getOpponent()
					.newInstance();
			final ScorePlayer score = new ScorePlayer(player1, player2, false);
			setCost(score.score());
		} catch (final InstantiationException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (final IllegalAccessException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

	}

}

File: CSIntroNeuralNetworkEdition2/Chapter6TicTacToe/Players/Neural/TicTacToeChromosome.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Reflection;

using HeatonResearchNeural.Feedforward;
using HeatonResearchNeural.Feedforward.Train.Genetic;
using Chapter6TicTacToe.Game;

namespace Chapter6TicTacToe.Players.Neural
{
    class TicTacToeChromosome : NeuralChromosome
    {
        /**
 * The constructor, takes a list of cities to set the initial "genes" to.
 * 
 * @param cities
 *            The order that this chromosome would visit the cities. These
 *            cities can be thought of as the genes of this chromosome.
 * @throws NeuralNetworkException
 */
        public TicTacToeChromosome(TicTacToeGenetic genetic,
                 FeedforwardNetwork network)
        {
            this.GA = genetic;
            this.Network = network;

            InitGenes(network.MatrixSize);
            UpdateGenes();
        }


        override public void CalculateCost()
        {


            // update the network with the new gene values
            this.UpdateNetwork();

            PlayerNeural player1 = new PlayerNeural(this.Network);
            Player player2;

            player2 = (Player)Assembly.GetExecutingAssembly().CreateInstance(this.getTicTacToeGenetic().getOpponent().FullName);

            ScorePlayer score = new ScorePlayer(player1, player2, false);

            double cost = score.score();
            // sometimes the score will be zero, just due to "luck"
            // if this is the case, try once more
            if (cost < 0.00001)
            {
                cost = score.score();
            }

            this.Cost = cost;
        }

        public TicTacToeGenetic getTicTacToeGenetic()
        {
            return (TicTacToeGenetic)this.GA;
        }

    }
}
